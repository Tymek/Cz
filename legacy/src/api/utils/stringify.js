/**
 * @see https://stackoverflow.com/a/48254637/1729641
 * @param {Object} v data
 * @param {string|number} space object that's used to insert white space
 */
const customStringify = (v, space = null) => {
  const cache = new Set()
  return JSON.stringify(v, (key, value) => {
    if (typeof value === 'object' && value !== null) {
      if (cache.has(value)) {
        return undefined
      }
      cache.add(value)
    }
    return value
  }, space)
}

export default customStringify
