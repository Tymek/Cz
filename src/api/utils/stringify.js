/**
 * @see https://stackoverflow.com/a/48254637/1729641
 * @param {Object} v data
 */
const customStringify = v => {
  const cache = new Set()
  return JSON.stringify(v, (key, value) => {
    if (typeof value === 'object' && value !== null) {
      if (cache.has(value)) {
        return undefined
      }
      cache.add(value)
    }
    return value
  })
}

export default customStringify
