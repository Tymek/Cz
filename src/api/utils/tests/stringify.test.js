import stringify from '../stringify'

describe('api/utils/stringify', () => {
  it('serializes JSON', () => {
    const input = { a: 1, b: 'test', test: 5.5 }
    return expect(stringify(input)).toEqual(JSON.stringify(input))
  })
  it('serializes curcular structure', () => {
    const input = {
      a: 'foo',
    }
    input.b = input
    return expect(stringify(input)).toMatchSnapshot()
  })
  it('serializes deep curcular structure', () => {
    const input = {
      id: 201,
      box: {
        owner: null,
        key: 'storm',
      },
      lines: [
        'item1',
        23,
      ],
    }

    input.box.input = input
    input.lines.box = input.box

    return expect(stringify(input)).toMatchSnapshot()
  })
})
