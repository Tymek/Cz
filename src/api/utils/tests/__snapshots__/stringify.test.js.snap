// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/utils/stringify serializes curcular structure 1`] = `"{\\"a\\":\\"foo\\"}"`;

exports[`api/utils/stringify serializes deep curcular structure 1`] = `"{\\"id\\":201,\\"box\\":{\\"owner\\":null,\\"key\\":\\"storm\\"},\\"lines\\":[\\"item1\\",23]}"`;
